fl4health/strategies/fedpca.py

```diff
@@ -0,0 +1,269 @@
+            merged_singular_vectors, merged_singular_values = self.merge_subspaces_svd(
+                client_singular_vectors, client_singular_values
+            )
+        else:
+            # use qr merging instead
+            log(INFO, "Performing QR-based merging.")
+            merged_singular_vectors, merged_singular_values = self.merge_subspaces_qr(
+                client_singular_vectors, client_singular_values
+            )
+        parameters = ndarrays_to_parameters([merged_singular_vectors, merged_singular_values])
+
+        # Aggregate custom metrics if aggregation fn was provided
+        metrics_aggregated = {}
+        if self.fit_metrics_aggregation_fn:
+            fit_metrics = [(res.num_examples, res.metrics) for _, res in results]
+            metrics_aggregated = self.fit_metrics_aggregation_fn(fit_metrics)
+        elif server_round == 1:  # Only log this warning once
+            log(WARNING, "No fit_metrics_aggregation_fn provided")
+
+        return parameters, metrics_aggregated
+
+    def merge_subspaces_svd(
+        self, client_singular_vectors: NDArrays, client_singular_values: NDArrays
+    ) -> Tuple[NDArray, NDArray]:
+        """
+        Produce the principal components (PCs) for all the data distributed across clients by merging the PCs
+        belonging to each local dataset.
+
+        Each clients sends a matrix whose columns are the local principal components to the server. The corresponding
+        singular values are also shared.
+
+        The server then arranges the local PCs into a block matrix, then perform SVD.
+
+        For example, if U_i denotes the matrix of PCs and S_i denotes the diagonal matrix of
+        singular values for client i, and there are n clients, then merging is done by
+        performing SVD on the matrix
+
+        B = [U_1 @ S_1 | U_2 @ S_2 | ... | U_n @ S_n],
+
+        where the new (left) singular vectors are returned as the merging result.
+
+        This implementation assumes that the *columns* of U_i are the PCs.
+
+        If
+
+        U @ S @ V.T = B
+
+        is the SVD of B, then it turns out that U = A @ U',
+        where the columns of U' are the true principal components of the aggregated data,
+        and A is some block unitary matrix.
+
+        For the theoretical justification behind this procedure, see the paper
+        "A Distributed and Incremental SVD Algorithm for Agglomerative Data Analysis on Large Networks".
+
+        Args:
+            client_singular_vectors (NDArrays): Local PCs.
+            client_singular_values (NDArrays): Singular values corresponding to local PCs.
+
+        Returns:
+            Tuple[NDArray, NDArray]: merged PCs and corresponding singular values.
+        """
+        X = [U @ np.diag(S) for U, S in zip(client_singular_vectors, client_singular_values)]
+        svd_input = np.concatenate(X, axis=1)
+        new_singular_vectors, new_singular_values, _ = np.linalg.svd(svd_input, full_matrices=True)
```

Comment:
If I recall correctly, the projection procedure is slightly different for right vs. left singular vectors. So we definitely want to make sure the math here is fully aligned.